tests/test_headers.py

```diff
@@ -117,6 +117,16 @@ def on_head(self, req, resp):
         resp.content_location = self.URL1
 
 
+class UnicodeHeaderResource:
+
+    def on_get(self, req, resp):
+        resp.set_headers([
+            (u'X-auTH-toKEN', 'toomanysecrets'),
+            ('Content-TYpE', u'application/json'),
+            (u'X-symBOl', u'\u0040'),
```

Comment:
Ok, adding a negative test shows different behaviors for py2k and py3k, namely:

```python
str(u"\U0001F37A").lower()    # beer
```

This will raise UnicodeEncodeError in py2k but is handled fine in py3k.

I've been going over RFC2616, and the relevant section is Section 4.2 (Message Headers) which references RFC822 section 3.1 (General Description). Best as I can tell, HTTP headers can be encoded in ISO-8859-1/Latin-1.

Not sure how we'll want to proceed. I know we chose str() because it's more performant than type-checking, but I'm not sure how to get consistent behavior without resorting to that.